frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx

```diff
@@ -32,37 +36,81 @@ const resource = {
 };
 
 const CloudShellTerminal: React.FC<CloudShellTerminalProps> = ({ username, onCancel }) => {
-  const [data, loaded, loadError] = useK8sWatchResource<CloudShellResource>(resource);
+  const [data, loaded, loadError] = useK8sWatchResource<CloudShellResource[]>(resource);
+  const [initData, setInitData] = React.useState<TerminalInitData>();
+  const [initDataLoading, setInitDataLoading] = React.useState<boolean>(false);
+  const [initError, setInitError] = React.useState<string>();
+  const [workspaceNamespace, setWorkspaceNamespace] = React.useState<string>();
 
-  if (loadError) {
+  React.useEffect(() => {
+    let destroy = false;
+    if (Array.isArray(data)) {
+      const workspace = data.find(
+        (ws) => ws?.metadata?.annotations?.[CLOUD_SHELL_USER_ANNOTATION] === username,
+      );
+      const running = workspace?.status?.phase === 'Running';
+
+      if (running) {
+        setInitDataLoading(true);
+        const { name, namespace } = workspace.metadata;
+        initTerminal(username, name, namespace)
+          .then((res: TerminalInitData) => {
+            if (destroy) return;
+            setInitData(res);
+            setInitDataLoading(false);
+            setWorkspaceNamespace(namespace);
+          })
+          .catch(() => {
+            if (destroy) return;
+            setInitDataLoading(false);
+            setInitError('Failed to connect to your OpenShift command line terminal');
+          });
+      }
+    }
+
+    return () => {
+      destroy = true;
+    };
+  }, [data, username]);
+
+  if (loadError || initError) {
     return (
-      <StatusBox loaded={loaded} loadError={loadError} label="OpenShift command line terminal" />
+      <StatusBox
+        loaded={loaded}
+        loadError={loadError || initError}
+        label="OpenShift command line terminal"
+      />
     );
   }
 
-  if (!loaded) {
-    return <LoadingBox />;
+  if (!loaded || initDataLoading) {
```

Comment:
Sometime between data being loaded, but not having run the effect to make the call to init, we skip this if condition resulting in a flash of the `CloudShellSetup` component.

We need to show the `CloudShellSetup` if data is loaded and the workspace is not found. For this reason finding the workspace in the data should probably be moved out of the effect.